### ISSUE TITLE
CIFAR10 accuracy for the quantized model is not increasing

### ISSUE THREAD

**Issue by joyalbin** [EunhyeokPark/PROFIT#1]
Hi.
I would like to thank for share this wonderful method for quantization.
I was just trying to experiment this solution with cifar10 dataset. I had replaced the imagenet data dependencies with CIFAR10.
But infact the accuracy for the quantized models are not showing as expected.

```
==> Full precision fine-tuning
Epoch: [14][500/521]>-Time 0.150 (0.163)>-Data 0.000 (0.002)>-Loss 1.0012 (1.0747)>-Prec@1 72.917 (70.056)>-Prec@5 93.750 (95.990)
 * Prec@1 90.000 Prec@5 100.000                                                 
 * Prec@1 90.000 Prec@5 100.000                                                 

==> Activation quantization, bit 8                                              
Epoch: [14][500/521]>-Time 0.216 (0.236)>-Data 0.000 (0.003)>-Loss 2.2443 (2.2571)>-Prec@1 17.708 (13.527)>-Prec@5 65.625 (62.169)
 * Prec@1 40.000 Prec@5 42.000                                                  
 * Prec@1 0.000 Prec@5 0.000 

==> Activation quantization, bit 5                                              
Epoch: [4][500/521]>Time 0.216 (0.236)>-Data 0.000 (0.002)>-Loss 2.2890 (2.2857)>-Prec@1 15.625 (12.227)>-Prec@5 55.208 (58.692)
 * Prec@1 2.000 Prec@5 48.000                                                   
 * Prec@1 100.000 Prec@5 100.000

==> Activation quantization, bit 4                                              
Epoch: [4][500/521]>Time 0.221 (0.239)>-Data 0.000 (0.002)>-Loss 2.3129 (2.2628)>-Prec@1 8.333 (13.519)>Prec@5 43.750 (60.754)
 * Prec@1 0.000 Prec@5 100.000                                                  
 * Prec@1 36.000 Prec@5 40.000

==> Weight quantization, bit 8
Epoch: [4][500/521]>Time 0.219 (0.245)>-Data 0.000 (0.002)>-Loss 2.3020 (2.3049)>-Prec@1 9.375 (10.004)>Prec@5 45.833 (49.885)
 * Prec@1 0.000 Prec@5 100.000                                                  
 * Prec@1 0.000 Prec@5 0.000 
```

floating point (32bit) accuracy for both _model and model_ema_ coming 90%
But the 8bit weight quantization accuracy is 0%

Is there anything else I have to takecare other than changing "num_class" and the pretrained model with CIFAR10 data? and I have changed "--teacher=self" in the training command
Could you please help me to figure out the issue?

Thanks in advance
Albin

**Comment by EunhyeokPark** [comment]
Hi, 

Could you let me know the model you used? 

The hyper-parameters of the models in this repository are tuned for the imagenet dataset. Thus they are not appropriate for CIFAR dataset. I think you used a different model I have, and I couldn't give you an answer because I don't know the details. 

In addition, it would be better to use longer epoch, or you should set the warm-up epochs appropriately in the scheduler.

**Comment by joyalbin** [comment]
@EunhyeokPark 
I am working with mobilnetV2
Sorry, I did a mistake while raising the issue. I was using using CIFAR10 dataset, not CIFAR100

I have used a pretrained model from a private https://github.com/huyvnphan/PyTorch_CIFAR10
https://drive.google.com/file/d/11DDSbPqFXLzooIv6YPmXuKRIZJ24808g/view
`mobilenet_v2.pt`

My FP32 training got 90% accuracy. The hyper parameters you meant for the progressive activation/weight training?

**Comment by EunhyeokPark** [comment]
Oh, I see...yes. I was a little curious about that, too. In general, if you train your model from scratch, MobileNetv2@CIFAR-100 shows about 75 % top-1 accuracy.

What I meant is a bn averaging rate, stride, etc. If you use a model for the CIFAR dataset, then it might be fine. 

Well, I figured out that the loss of your 8-bit activation quantization is too big and the accuracy is too low.  I think it's already broken at that stage. In order to apply quantization, you should insert quantization layers appropriately into your model, and you can find an example at PROFIT/conv_model/ilsvrc/.  In the case of MobileNet-v2, you should use different quantization layers depending on the activation distribution, i.e., symmetric or non-negative only.

**Comment by joyalbin** [comment]
@EunhyeokPark 
Do you mean, mobilnetv2 with even different dataset, the CONVILUTION and other operators like ReLU, the quantization logic may need change?

Quantization logic:
```
    def initialize(self, n_lv):                                                 
        self.n_lv = n_lv                                                        
        max_val = self.weight.data.abs().max().item()                           
        self.a.data.fill_(np.log(np.exp(max_val * 0.9)-1))                      
        self.c.data.fill_(np.log(np.exp(max_val * 0.9)-1))                      
                                                                                
    def _weight_quant(self):                                                    
        a = F.softplus(self.a)                                                  
        c = F.softplus(self.c)                                                  
                                                                                
        weight = F.hardtanh(self.weight / a, -1, 1)                             
        weight = RoundQuant.apply(weight, self.n_lv // 2) * c                   
        return weight
```

**Comment by EunhyeokPark** [comment]
In PROFIT/conv_model/ilsvrc/MobileNetv2_quant.py, you can see that ops.ReLU6 and ops.Sym layer are inserted. These layers are responsible for activation quantization, but each of them has a different quantization range so they should be selected carefully based on the target distribution. You need to modify your MobileNet-v2 model for CIFAR dataset in this way. 

Indeed, I think you already applied some of the quantization layers to your code. Otherwise, there should be no accuracy degradation. However, the quality of the quantized model is highly sensitive to the quantization range, so please check your code carefully.

**Comment by joyalbin** [comment]
ok, let me rework on that. Thank you for your valuable reply.
